src/libraries/System.Text.Json/src/System/Text/Json/Serialization/Converters/JsonDictionaryDefaultConverter.cs

```diff
@@ -0,0 +1,311 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+
+namespace System.Text.Json.Serialization.Converters
+{
+    /// <summary>
+    /// Default base class implementation of <cref>JsonDictionaryConverter{TCollection}</cref> .
+    /// </summary>
+    internal abstract class JsonDictionaryDefaultConverter<TCollection, TValue>
+        : JsonDictionaryConverter<TCollection>
+    {
+        /// <summary>
+        /// When overridden, adds the value to the collection.
+        /// </summary>
+        protected abstract void Add(TValue value, JsonSerializerOptions options, ref ReadStack state);
+
+        /// <summary>
+        /// When overridden, converts the temporary collection held in state.Current.ReturnValue to the final collection.
+        /// This is used with immutable collections.
+        /// </summary>
+        protected virtual void ConvertCollection(ref ReadStack state, JsonSerializerOptions options) { }
+
+        /// <summary>
+        /// When overridden, create the collection. It may be a temporary collection or the final collection.
+        /// </summary>
+        protected virtual void CreateCollection(ref ReadStack state) { }
+
+        internal override Type ElementType => typeof(TValue);
+
+        protected static JsonConverter<TValue> GetElementConverter(ref ReadStack state)
+        {
+            JsonConverter<TValue>? converter = state.Current.JsonClassInfo.ElementClassInfo!.PolicyProperty!.ConverterBase as JsonConverter<TValue>;
+            if (converter == null)
+            {
+                state.Current.JsonClassInfo.ElementClassInfo.PolicyProperty.ThrowCollectionNotSupportedException();
+            }
+
+            return converter!;
+        }
+
+        protected string GetKeyName(string key, ref WriteStack state, JsonSerializerOptions options)
+        {
+            if (options.DictionaryKeyPolicy != null && !state.Current.IgnoreDictionaryKeyPolicy)
+            {
+                key = options.DictionaryKeyPolicy.ConvertName(key);
+
+                if (key == null)
+                {
+                    ThrowHelper.ThrowInvalidOperationException_SerializerDictionaryKeyNull(options.DictionaryKeyPolicy.GetType());
+                }
+            }
+
+            return key;
+        }
+
+        protected JsonConverter<TValue> GetValueConverter(ref WriteStack state)
+        {
+            JsonConverter<TValue> converter = (JsonConverter<TValue>)state.Current.DeclaredJsonPropertyInfo.ConverterBase;
```

Comment:
Why this method uses an explicit cast while `GetElementConverter` uses the `as` operator?

Reply:
No `as` call is necessary here. Removing.

Reply:
Explicit cast should always work; removing `as`